go/vt/wrangler/reparent.go

```diff
@@ -808,37 +837,71 @@ func (wr *Wrangler) chooseNewMaster(
 	tabletMap map[string]*topo.TabletInfo,
 	avoidMasterTabletAlias *topodatapb.TabletAlias,
 	waitReplicasTimeout time.Duration) (*topodatapb.TabletAlias, error) {
-
-	if avoidMasterTabletAlias == nil {
-		return nil, fmt.Errorf("tablet to avoid for reparent is not provided, cannot choose new master")
+	candidates, err := wr.FindMasterCandidates(ctx, tabletMap, waitReplicasTimeout)
+	if err != nil {
+		return nil, err
 	}
+
 	var masterCell string
 	if shardInfo.MasterAlias != nil {
 		masterCell = shardInfo.MasterAlias.Cell
 	}
 
-	maxPosSearch := maxReplPosSearch{
+	// Filter out avoidMasterTabletAlias, if it exists in the candidates list, and also filter out
+	// any candidates that aren't in the master tablets cell.
+	filteredCandidates := make([]*topodatapb.TabletAlias, 0, len(candidates))
+	for i := range candidates {
+		if candidates[i] != avoidMasterTabletAlias && candidates[i].Cell == masterCell {
+			filteredCandidates = append(filteredCandidates, candidates[i])
+		}
+	}
+
+	if len(filteredCandidates) == 0 {
+		return nil, nil
+	}
+	return filteredCandidates[0], nil
+}
+
+// FindMasterCandidates will look at all of the tablets in the supplied tabletMap, and return a list of tabletAliases that
+// are all caught up. This means that all of the returned TabletAliases will have the same replication position.
```

Comment:
If I'm reading this right, I don't see a guarantee (or a check) that the returned replicas are actually caught up. So we should fix the comments: "all caught up" -> "the most up-to-date" OR "the most caught up".

Reply:
Since we ignore tablets we can't reach, we should clarify that it's additionally "the most up-to-date that we were able to reach". It's therefore important to still call `CheckTabletIsFarthestAhead`.